network/p2p/gossip/gossip.go

```diff
@@ -231,84 +237,190 @@ func (p *PullGossiper[_]) handleResponse(
-			_, _ = p.pending.PopLeft()
+			delete(p.tracking, gossipID)
 			return err
 		}
 
 		gossip = append(gossip, bytes)
 		sentBytes += len(bytes)
-		p.pending.PopLeft()
+		p.issued.PushRight(gossipable)
+		p.tracking[gossipID] = now
 	}
 
+	maxLastGossipTimeToRegossip := now.Add(-p.maxRegossipFrequency)
+
+	// Iterate over all issued gossipables (have been sent before) to fill any
+	// remaining space in gossip batch.
+	for sentBytes < p.targetGossipSize {
+		gossipable, ok := p.issued.PopLeft()
+		if !ok {
+			break
+		}
+
+		// Ensure item is still in the set before we gossip.
+		gossipID := gossipable.GossipID()
+		if !p.set.Has(gossipID) {
+			delete(p.tracking, gossipID)
+			p.discarded.Put(gossipID, nil) // only add to discarded if issued once
+			continue
+		}
+
+		// Ensure we don't attempt to send a gossipable too frequently.
+		lastGossipTime := p.tracking[gossipID]
+		if maxLastGossipTimeToRegossip.Before(lastGossipTime) {
+			// Put the gossipable on the front of the queue to keep items sorted
+			// by last issuance time.
+			p.issued.PushLeft(gossipable)
+			break
+		}
+
+		bytes, err := p.marshaller.MarshalGossip(gossipable)
+		if err != nil {
+			// Should never happen because we've already issued this once.
+			delete(p.tracking, gossipID)
+			return err
+		}
+
+		gossip = append(gossip, bytes)
+		sentBytes += len(bytes)
+		p.issued.PushRight(gossipable)
+		p.tracking[gossipID] = now
+	}
+	p.metrics.tracking.Set(float64(len(p.tracking)))
+
+	// If there is nothing to gossip, we can exit early.
+	if len(gossip) == 0 {
+		return nil
+	}
+
+	// Send gossipables to peers
 	msgBytes, err := MarshalAppGossip(gossip)
 	if err != nil {
 		return err
 	}
-
 	sentCountMetric, err := p.metrics.sentCount.GetMetricWith(pushLabels)
 	if err != nil {
 		return fmt.Errorf("failed to get sent count metric: %w", err)
 	}
-
 	sentBytesMetric, err := p.metrics.sentBytes.GetMetricWith(pushLabels)
 	if err != nil {
 		return fmt.Errorf("failed to get sent bytes metric: %w", err)
 	}
-
 	sentCountMetric.Add(float64(len(gossip)))
 	sentBytesMetric.Add(float64(sentBytes))
-
-	return p.client.AppGossip(ctx, msgBytes)
+	if err := p.client.AppGossip(ctx, msgBytes); err != nil {
+		return fmt.Errorf("failed to gossip: %w", err)
+	}
+	return nil
 }
 
+// Add enqueues new gossipables to be pushed. If a gossiable is already tracked,
+// it is not added again.
 func (p *PushGossiper[T]) Add(gossipables ...T) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
+	// Add new gossipables to the pending queue.
+	now := time.Now()
 	for _, gossipable := range gossipables {
-		p.pending.PushRight(gossipable)
+		gossipID := gossipable.GossipID()
+		if _, ok := p.tracking[gossipID]; ok {
+			continue
+		}
+		if _, ok := p.discarded.Get(gossipID); ok {
+			// Pretend that recently discarded transactions were just gossiped.
+			p.tracking[gossipID] = now
+			p.issued.PushRight(gossipable)
+		} else {
+			p.tracking[gossipID] = time.Time{}
```

Comment:
why do we use `time.Time{}` instead of now if it was not in the discarded cache?

Reply:
We treat something in the discarded cache as having been gossiped at the time of issuance. Otherwise, we mark it as having never been gossiped.